src/BuiltInTools/dotnet-watch/Internal/FileWatcher.cs

```diff
@@ -129,14 +99,39 @@ private void EnsureNotDisposed()
         }
 
         private static string EnsureTrailingSlash(string path)
+            => (path is [.., var last] && last != Path.DirectorySeparatorChar) ? path + Path.DirectorySeparatorChar : path;
+
+        public async Task<FileItem?> GetChangedFileAsync(Action? startedWatching, CancellationToken cancellationToken)
         {
-            if (!string.IsNullOrEmpty(path) &&
-                path[path.Length - 1] != Path.DirectorySeparatorChar)
+            StartWatching();
+
+            var fileChangedSource = new TaskCompletionSource<FileItem?>(TaskCreationOptions.RunContinuationsAsynchronously);
+            cancellationToken.Register(() => fileChangedSource.TrySetResult(null));
+
+            void FileChangedCallback(string path, bool newFile)
             {
-                return path + Path.DirectorySeparatorChar;
+                if (fileSet.TryGetValue(path, out var fileItem))
+                {
+                    fileChangedSource.TrySetResult(fileItem);
+                }
             }
 
-            return path;
+            OnFileChange += FileChangedCallback;
+            startedWatching?.Invoke();
+            var changedFile = await fileChangedSource.Task;
+            OnFileChange -= FileChangedCallback;
```

Comment:
Do you want the -= FileChangedCallback in a try/finally, in case fileChangedSource.Task throws/is cancelled?

Reply:
Yup.